### ISSUE TITLE
Optimised dequeing

### ISSUE THREAD

**Issue by mohmad-null** [scrapy/scrapy#2474]
A feature request.
I've got my settings set up so that Scrapy is being nice and polite to any given site. For example:

```
# Very high
CONCURRENT_REQUESTS = 128

CONCURRENT_REQUESTS_PER_IP = 3
DOWNLOAD_DELAY = 2

AUTOTHROTTLE_ENABLED = True
AUTOTHROTTLE_TARGET_CONCURRENCY = 3
AUTOTHROTTLE_START_DELAY = 1
AUTOTHROTTLE_MAX_DELAY = 15

#because I want to try and focus on breadth - I'm mostly making few a requests to many many domains
#With a few domains getting many requests
DEPTH_PRIORITY = 1
SCHEDULER_DISK_QUEUE = 'scrapy.squeues.PickleFifoDiskQueue'
SCHEDULER_MEMORY_QUEUE = 'scrapy.squeues.FifoMemoryQueue'
```

As you can see, for any given IP, scrapy is going to be slow, but in theory scrapy can handle many different IP's at once - that's why CONCURRENT_REQUESTS is very high. The problem is, it doesn't seem to.

Consider the following hypothetical queue (assume all domains are on different IPs):

> example.com/1
> example.com/2
> example.com/3
> example.com/4
> example.com/5
> example.com/6
> example.com/7
> example.com/8
> google.example/1
> google.example/2
> google.example/3
> google.example/4
> google.example/5
> google.example/6
> google.example/7
> blue.example
> yellow.example
> red.example
> green.example
> purple.example
> pink.example
> blank.example

Assuming scrapy is parsing the above top-down (and ignoring whether we're doing FIFO or reverse), what seems to happen is that scrapy will scrape all of the example.com pages first, then the google.example ones, in both instances hitting the per-ip/autothrottle limits and waiting for them to clear down.

However, the optimal way to do the above list is to note all the different domains, and then try and do them all concurrently (within the limit of CONCURRENT_REQUESTS). This doesn't appear to happen at present.

(At least, I'm guessing that's the reason I'm seeing very slow crawl rates, a low number of tcp/ip threads, as well as negligible CPU usage (<20% of a single core), and almost non-existent network usage (about 1MB per min - 130k bits/s!!!))

**Comment by kmike** [comment]
Yeah, this is a problem we need to solve. See also https://github.com/scrapy/scrapy/issues/2435 and https://github.com/scrapy/scrapy/issues/1802 - there are some workarounds.

**Comment by mohmad-null** [comment]
Interesting, thanks for the links.

In lieu of a fancy new algorithm, I think my suggestion would be for a random de-queuing feature.

So rather than FIFO or LIFO, it randomly picks an item from the queue and crawls that instead assuming it's allowed (i.e. autothrottle etc taken into account). 
For bonus points, if it can't do that url (i.e. autothrottle), it could pop it back into the queue and try another one up to n times before just waiting for autothrottle or whatever to allow it to pass.

I'd imagine at least the first part of that would be relatively simple to implement (although my attempt to grok that portion of the code didn't get me far), much more so than a new featureful algorithm, and for use-cases like mine (~3-10 requests to each of thousands of domains, it would probably work almost as well as a fancy algorithm).

**Comment by kmike** [comment]
@mohmad-null example implementation: https://github.com/TeamHG-Memex/linkdepth/blob/master/queues.py

To use it put 'scheduler_slot' to request.meta and add the following options to scrapy settings.py:

```
SCHEDULER_PRIORITY_QUEUE = 'queues.RoundRobinPriorityQueue'

# optional, only if you're using disk queues
# it has caveats - crawl resume still doesn't work,
# the only advantage is saved RAM
SCHEDULER_DISK_QUEUE = 'queues.DiskQueue' 

```

**Comment by milescrawford** [comment]
Chiming in to say this would be a very important feature to us.  

Random/round-robin dequeue is barely a mitigation, because, when a site with many URIs in the frontier is waiting in a few slots, you slowly round-robin more and more URIs from that site into the slots. Other sites are handled quickly, the dominant site is handled slowly and thus progressively captures more of the slots over time.

Frontera is a very different product with vastly greater operational overhead, and while it is an option, it's really a whole new approach.

**Comment by kmike** [comment]
@milescrawford thanks for chiming in, it helps to see use cases!
I'm not sure I understand your problem though: round-robin is for choosing a domain, so you're getting the same amount of requests for each domain, regardless of a total number of requests for a domain in a queue. If there are no domains left then it is most efficient to crawl this single remaining domain, and if there are other domains left they'll be selected. It seems I'm missing something.

Another data point in [deep-deep](https://github.com/TeamHG-Memex/deep-deep) crawler: we used a queue which first chooses a domain to crawl, with a probability which depends on a maximum priority of a request for this domain. Implementation: https://github.com/TeamHG-Memex/deep-deep/blob/master/deep-deep/deepdeep/queues.py. 

+1 to have a built-in solution in Scrapy, and making sure it is useful for Frontera (at least APIs).

**Comment by milescrawford** [comment]
This arises particularly if the dominant site is slower or has a longer crawl delay than the others. 

Each domain gets the same amount of requests for round-robin or a proportional amount if random.

However, if a previous request for the dominant site hasn't cleared by the time it round-robins up again, now two slots are dedicated to it. The second slot now has to wait the delay before it can proceed, during which time another one arrives, etc. It snowballs until that site has all or a majority of the slots, reducing how fast you can move through the other sites as they round-robin up.

**Comment by kmike** [comment]
@milescrawford got it now, thanks for the explanation! 

It seems that frontera's approach could work better for this use case: instead of randomizing domains or selecting them in a round-robin fashion frontera's scheduler asks scrapy downloader which slots are busy, and avoids adding requests to them.

**Comment by milescrawford** [comment]
Yes, I'm considering making a channel by which my frontier implementation can see what domains are in progress and choose accordingly.

**Comment by milescrawford** [comment]
I put the function below into a bit of middleware so that I get a continuously updated list of dominant sites currently in progress.

Now my frontier knows which sites to avoid and can filter them out when dequeuing.  This provided me with a 20x gain in throughput!

```
from collections import Counter

# Sites occupying more than this fraction of the current work won't be dequeued.
DOMINANT_SITE_THRESHOLD = 0.05

def _update_dominant_sites(self, spider):
    """Inspect the active slots let the frontier know if there are any over-represented sites to avoid."""
    inprogress = spider.crawler.engine.slot.inprogress
    sites = list(map(lambda request: urllib.parse.urlparse(request.url).netloc, inprogress))
    maximum = spider.crawler.settings.get('CONCURRENT_REQUESTS')
    per_site = Counter(sites)
    dominant = {site: count for (site, count) in per_site.items() if count / maximum > DOMINANT_SITE_THRESHOLD}
    frontier.set_dominant_sites(list(dominant.keys()))
```

**Comment by tianhuil** [comment]
I would like to propose adding a `DomainScheduler` [implemented here](https://github.com/tianhuil/domain_scheduler) into scrapy.  It scrapes in a domain-smart way: by round-robin cycling through the domains. This has two benefits:

1. Spreading out load on the target server instead of hitting the server with many requests at once
2. Reducing delays caused by server-overloaded errors or CONCURRENT_REQUESTS_PER_IP type restrictions.

It implements the solution proposed in https://github.com/scrapy/scrapy/issues/1802#issue-135260562.  It is dependent on https://github.com/scrapy/queuelib/pull/21.  Thoughts?

**Comment by kmike** [comment]
This is fixed in general by #3520.
